driver_test.go

```diff
@@ -12,26 +12,246 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package spannerdriver_test
+package spannerdriver
 
 import (
+	"cloud.google.com/go/spanner"
+	"context"
 	"database/sql"
+	"fmt"
 	"log"
+	"os"
+	"reflect"
+	"runtime/debug"
+	"testing"
 
-	spannerdriver "github.com/rakyll/go-sql-driver-spanner"
+	// api/lib packages not imported by driver
+	adminapi "cloud.google.com/go/spanner/admin/database/apiv1"
 	"google.golang.org/api/option"
+	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+	"google.golang.org/grpc"
 )
 
-func ExampleDriver() {
-	driver := &spannerdriver.Driver{
-		Options: []option.ClientOption{
-			option.WithCredentialsFile("/path/to/service-account-key.json"),
-		},
+var (
+	project  string
+	instance string
+	dbid   string
+	dsn      string
+	emhost	string
+	projset  bool
+	instset  bool
+	dbset    bool
+	emset	bool
+)
+
+type Connector struct {
+	ctx         context.Context
+	client      *spanner.Client
+	adminClient *adminapi.DatabaseAdminClient
+}
+
+func NewConnector(emhost string, emset bool) (*Connector, error) {
+
+	ctx := context.Background()
+
+	var adminClient *adminapi.DatabaseAdminClient
+	var err error
+
+	if emset{
+		adminClient, err = adminapi.NewDatabaseAdminClient(
+			ctx,
+			option.WithoutAuthentication(),
+			option.WithEndpoint(emhost),
+			option.WithGRPCDialOption(grpc.WithInsecure()))
+		if err != nil {
+			return nil, err
+		}
+	}else{
+		adminClient, err = adminapi.NewDatabaseAdminClient(ctx)
+		if err != nil {
+			return nil, err
+		}
 	}
-	connector, err := driver.OpenConnector("projects/$PROJECT/instances/$INSTANCE/databases/$DATABASE")
+
+	dataClient, err := spanner.NewClient(ctx, dsn)
 	if err != nil {
+		return nil, err
+	}
+
+	curs := &Connector{
+		ctx:         ctx,
+		client:      dataClient,
+		adminClient: adminClient,
+	}
+	return curs, nil
+}
+
+func (c *Connector) Close() {
+	c.client.Close()
+	c.adminClient.Close()
+}
+
+// structs for row data
+type testaRow struct {
+	A string
+	B string
+	C string
+}
+type typeTestaRow struct {
+	stringt string
+	intt    int
+	floatt  float64
+	boolt   bool
+}
+
+func init() {
+
+	// get environment variables or set to default
+	if instance, instset = os.LookupEnv("SPANNER_TEST_INSTANCE"); !instset {
+		instance = "test-instance"
+	}
+	if project, projset = os.LookupEnv("SPANNER_TEST_PROJECT"); !projset {
+		project = "test-project"
+	}
+	if dbid, dbset = os.LookupEnv("SPANNER_TEST_DBID"); !dbset {
+		dbid = "gotest"
+	}
+
+	// derive data source name
+	dsn = "projects/" + project + "/instances/" + instance + "/databases/" + dbid
+
+	// configure emulator if set 
+	emhost, emset = os.LookupEnv("SPANNER_EMULATOR_HOST")
```

Comment:
nit: this should be `spanner_host`, instead of `emhost`, since the host could be a non emulator one.